packages/agent/test/agent_spec.ts

```diff
@@ -0,0 +1,181 @@
+  })
+
+  after(function() {
+    return this.servers.stop()
+  })
+
+  ;[
+    {
+      name: 'With no upstream',
+    },
+    {
+      name: 'With an HTTP upstream',
+      proxyUrl: `http://localhost:${PROXY_PORT}`,
+    },
+    {
+      name: 'With an HTTPS upstream',
+      proxyUrl: `https://localhost:${PROXY_PORT}`,
+      httpsProxy: true,
+    },
+    {
+      name: 'With an HTTP upstream requiring auth',
+      proxyUrl: `http://foo:bar@localhost:${PROXY_PORT}`,
+      proxyAuth: true,
+    },
+    {
+      name: 'With an HTTPS upstream requiring auth',
+      proxyUrl: `https://foo:bar@localhost:${PROXY_PORT}`,
+      httpsProxy: true,
+      proxyAuth: true
+    }
+  ].slice().map((testCase) => {
+    context(testCase.name, function() {
+      beforeEach(function() {
+        this.oldEnv = Object.assign({}, process.env)
+        process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0'
+
+        if (testCase.proxyUrl) {
+          process.env.HTTP_PROXY = process.env.HTTPS_PROXY = testCase.proxyUrl
+          process.env.NO_PROXY = ''
+        }
+
+        this.agent = new CombinedAgent()
+
+        this.request = request.defaults({
+          proxy: null,
+          strictSSL: false,
+          agent: this.agent
+        })
+
+        if (testCase.proxyUrl) {
+          let options: any = {
+            keepRequests: true,
+            https: false,
+            auth: false
+          }
+
+          if (testCase.httpsProxy) {
+            options.https = this.https
+          }
+
+          if (testCase.proxyAuth) {
+            options.auth = {
+              username: 'foo',
+              password: 'bar'
+            }
+          }
+
+          this.debugProxy = new DebuggingProxy(options)
+          return this.debugProxy.start(PROXY_PORT)
+        }
+      })
+
+      afterEach(function() {
+        process.env = this.oldEnv
+
+        if (testCase.proxyUrl) {
+          this.debugProxy.stop()
+        }
+      })
+
+      it('HTTP pages can be loaded', function() {
+        return this.request({
+          url: `http://localhost:${HTTP_PORT}/get`,
+        }).then(body => {
+          expect(body).to.eq('It worked!')
+          if (this.debugProxy) {
+            expect(this.debugProxy.requests[0]).to.include({
+              url: `http://localhost:${HTTP_PORT}/get`
+            })
+          }
+        })
+      })
+
+      it('HTTPS pages can be loaded', function() {
+        return this.request({
+          url: `https://localhost:${HTTPS_PORT}/get`
+        }).then(body => {
+          expect(body).to.eq('It worked!')
+          if (this.debugProxy) {
+            expect(this.debugProxy.requests[0]).to.include({
+              https: true,
+              url: `localhost:${HTTPS_PORT}`
+            })
+          }
+        })
+      })
+
+      it('HTTP errors are catchable', function() {
+        return this.request({
+          url: `http://localhost:${HTTP_PORT}/empty-response`,
+        }).catch(err => {
```

Comment:
there is no proof that this `.catch()` handler is ever being called...

anytime you're doing this you have to ensure that the `onSuccess` fulfillment handler (aka the `.then()`) isn't being called.

in other tests of ours we'll usually do something like...

```js
doTheThing()
.then(() => {
  throw new Error('promise should not have resolved but it did')
})
.catch((err) => {
  ...
})
```

Bluebird also has a `.throw()` and a `.thenThrow` which would save you the callback function.

There are of course other ways to write it. You could use a `(done)` callback but that can get messy since you have conditional logic in the `.catch` handler. Likely easiest to